src/cj.c

```diff
@@ -0,0 +1,126 @@
+#include "data.table.h"
+
+SEXP cj(SEXP base_list) {
+  int JJ = LENGTH(base_list), nprotect = 0;
+  SEXP out = PROTECT(allocVector(VECSXP, JJ)); nprotect++;
+  int NN = 1;
+  for (int j = 0; j < JJ; j++) NN *= LENGTH(VECTOR_ELT(base_list, j));
+  int div = NN, modulo;
+  
+  //#pragma omp parallel for num_threads(getDTthreads())
```

Comment:
generally better to parallelise upper layers if possible, thus parallelising columns should give better improvement

Reply:
See comment above... any suggestions for how to do that? I couldn't see how to make the parallelization work without substantially complicating the logic when there's string input (or list for that matter, but we could block lists if that's a real impediment)

Or maybe I'm just missing something...

Reply:
Also, I guess most of the time the input to `CJ` is one, two, or three columns, but output could still be millions of rows... doesn't it make more sense to parallelize across rows then?

Reply:
make sense!